### PR TITLE
M24H-20 generate case number in jsbox

### PR DESCRIPTION
**Pull request by smn**

### CODE REVIEW

File: go-app.js [hunk 1]
```diff
@@ -46,6 +48,27 @@ go.utils = {
         return ('' + sum).slice(-1) == check;
     },
 
+    now: function () {
+        return moment();
+    },
```
Comment (smn) [inline go-app.js:53]: Doing this so I can stub out the time in the tests, they have to match dates that are mostly hardcoded in the fixtures.

File: go-app.js [hunk 2]
```diff
@@ -46,6 +48,27 @@ go.utils = {
...
+            return result.value;
+        })
+        .then(function(sequence_number) {
+            return [
+                go.utils.now().format('YYYYMMDD'),
+                facility_code,
+                sequence_number,
+            ].join('-');
+        });
+    },
```
Comment (smn) [inline go-app.js:70]: `kv.incr` is an automic thing in Vumi Go, it's backed by Redis, essentially I'm incrementing a counter on a key that's namespaced using the facility code.

File: test/app.test.js [hunk 3]
```diff
@@ -1,33 +1,65 @@
...
+                            api_endpoint: 'http://www.example.org/api/v1/',
+                            landmarks: JSON.parse(
+                                fs.readFileSync(
+                                    "src/lookups/landmarks.json", "utf8"))
+                        })
+                        .setup(function(api) {
+                            fixtures().forEach(api.http.fixtures.add);
+                            onafixtures.store.forEach(function (fixture) {
+                                api.http.fixtures.add(fixture);
+                            });
```
Comment (smn) [inline test/app.test.js:61]: All these changes are to make the app aware of Ona and to load Ona fixtures as HTTP fixtures which we can introspect in the tests.

File: test/app.test.js [hunk 4]
```diff
@@ -372,10 +410,51 @@ describe("app", function() {
...
                     .check.reply.content(/Thank you! Your report has been submitted/)
                     .run();
             });
         });
+
+        describe('Submit_Case', function () {
+            it('should make a valid Ona API call', function () {
+                return tester
+                    .setup.user.state('Submit_Case')
+                    .setup.user.answers(completed_answers)
```
Comment (smn) [inline test/app.test.js:424]: Loading all the "answers" the user would've given in earlier states so we have a complete dictionary of stuff we'd be posting to Ona.

File: test/app.test.js [hunk 5]
```diff
@@ -372,10 +410,51 @@ describe("app", function() {
...
+                    .input('1')
+                    .check.interaction({
+                        state: 'Facility_Code_Entry',
+                        reply: /Welcome! To report a malaria case, please enter your facility code./
+                    })
+                    .check(function (api) {
+                        var http_sent = _.where(api.http.requests, {
+                            url: 'http://ona.io/api/v1/submission'
+                        })[0];
```
Comment (smn) [inline test/app.test.js:433]: Get the HTTP request that was made to Ona for which we're returning a fixture as a mock response.

File: test/app.test.js [hunk 6]
```diff
@@ -372,10 +410,51 @@ describe("app", function() {
...
+                                locality_other: 'Locality_Entry_Other',
+                                msisdn: 'MSISDN_Entry',
+                                date_of_birth: '890510',
+                                create_date_time: '2015-01-01T00:00:00+00:00',
+                                abroad: 'Patient_Abroad_Entry',
+                                sa_id_number: '8905100273087',
+                                gender: 'gender',
+                                landmark: 'Landmark_Entry',
+                                landmark_description: 'Landmark_Entry_Description',
+                                case_number: '20150101-111111-1'
```
Comment (smn) [inline test/app.test.js:452]: Making sure I'm posting what I think I should be posting.

**Comment by smn** [comment]
@KaitCrawford ready for review, this one's likely a bit more tricky. I've left some explanatory comments in it, hope it helps.

File: go-app.js (hunk 2 above)
Comment (KaitCrawford) [inline go-app.js]: Where did all of this code come from? I thought go-app.js was the file generated by grunt but I don't see this stuff in src/app.js below?

**Comment by KaitCrawford** [comment]
:+1: Thanks for the explanations :smile:

**Pull request merged**